src/compiler/checker.ts

```diff
@@ -21425,6 +21428,12 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         return !!((type.flags & TypeFlags.Union ? (type as UnionType).types[0] : type).flags & TypeFlags.Undefined);
     }
 
+
+    function containsNonMissingUndefinedType(type: Type) {
+        const candidate = type.flags & TypeFlags.Union ? (type as UnionType).types[0] : type;
+        return !!(candidate.flags & TypeFlags.Undefined) && candidate !== missingType;

```

Comment:
Does `missingType` always precede `undefinedType`? I actually thought the order was flipped.

Reply:
You can't have both in a union - if non-missing `undefined` is added to a union, `missing` gets removed.